src/icrar/leap-accelerate/core/logging.cc

```diff
@@ -20,48 +20,63 @@
  * MA 02111 - 1307  USA
  */
 
-#include "logging.h"
+#include <boost/log/attributes/mutable_constant.hpp>
+#include <boost/log/core.hpp>
+#include <boost/log/expressions.hpp>
+#include <boost/log/sinks/text_file_backend.hpp>
+#include <boost/log/sources/global_logger_storage.hpp>
+#include <boost/log/sources/record_ostream.hpp>
+#include <boost/log/sources/severity_channel_logger.hpp>
+#include <boost/log/sources/severity_logger.hpp>
 #include <boost/log/support/date_time.hpp>
+#include <boost/log/utility/manipulators/add_value.hpp>
+#include <boost/log/utility/setup/common_attributes.hpp>
+#include <boost/log/utility/setup/console.hpp>
+#include <boost/log/utility/setup/file.hpp>
+
+#include "logging.h"
```

Comment:
I think the fact that `logging.h` is included on its own (i.e., without the rest of these boost headers) in the rest of the files should be enough? I indeed took care that it was self-contained, which is the reason why I moved all these boost headers to the .cpp file from the .h header.

Reply:
You're right in that the header is currently self-contained, it's just convenient if it's own source file always throws an error if loses that status in future which is common when performing large refactoring.

Reply:
That could be a bit misleading actually: for example, the header could contain a macro that used a definition coming from the boost headers included in the `cc` file. So the macro would work in `logging.cc` even if `logging.h` was included earlier, but would still fail to compile when using it in other `cc` files. I think I'll be sticking to system/third-party/local sorting for this one, as self-containment should hold true regardless of ordering of include statements and compilation units -- it's just one of those things that is a bit difficult to prove.

However, thanks to your comment I noticed the file was being included simply as `logging.h` rather than with the full path. I think gcc was just being lenient here and allowing this to happily compile, but other compilers might not have been so generous, so I changed it to the proper full path.

Reply:
This is just what's stated in https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes, but I also slightly disagree with the full order written there as putting system headers first isn't required if https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#sf10-avoid-dependencies-on-implicitly-included-names and https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#sf11-header-files-should-be-self-contained are proactively being met. Putting system headers first however does work in more situations as it sets up transitive includes and handles cases when you include a header that is not self contained.

The most notable mixed up library I can think of is `gl/gl.h`, `gl/glew.h` and `gl/glut.h` where `glew.h` has custom error code in it stating in it that you must first include `gl.h` before `glew.h` in your source file.

Reply:
With the local includes every compiler always include the source file directory which is fine for this projects structure at the moment.

Some libraries have public headers split into src/ and include/ to prevent dependant apps from including .cc/.cpp files which may break these short local includes (not the most fun refactor). I recall there was some large codebase that bans use of local includes for this sort of reason, but https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#sf12-prefer-the-quoted-form-of-include-for-files-relative-to-the-including-file-and-the-angle-bracket-form-everywhere-else says otherwise.

Reply:
Wow, this is a total surprise to me, I didn't know about relative paths being considered for the `#include ""` form, thanks for the pointers on that!